lib/Feature/FeatureModelTransaction.cpp

```diff
@@ -36,6 +39,98 @@ void removeFeature(FeatureModel &FM,
   Trans.commit();
 }
 
+Feature *getActualFeature(FeatureModel &FM, detail::FeatureVariantTy &FV) {
+  Feature *ActualFeature = nullptr;
+  std::visit(
+      Overloaded{[&ActualFeature, &FM](Feature *F) { ActualFeature = F; },
+                 [&ActualFeature, &FM](string &FName) {
+                   ActualFeature = FM.getFeature(FName);
+                 }},
+      FV);
+  return ActualFeature;
+}
+
+std::optional<bool> fvIsLeave(FeatureModel &FM, detail::FeatureVariantTy &FV) {
+  Feature *ActualFeature = getActualFeature(FM, FV);
+  // if Feature does not exist in FM
+  if (ActualFeature == nullptr) {
+    return std::nullopt;
+  }
+  return std::optional<bool>{ActualFeature->isLeave()};
+}
+
+std::optional<bool>
+canBeDeletedRecursively(FeatureModel &FM, detail::FeatureVariantTy &FV,
+                        const std::set<FeatureTreeNode *> &OtherFeatures) {
+  Feature *ActualFeature = getActualFeature(FM, FV);
+  // if Feature does not exist in FM
+  if (ActualFeature == nullptr) {
+    return std::nullopt;
+  }
+
+  std::set<FeatureTreeNode *> Intersection;
+  auto AllChildrenInSubtree = ActualFeature->getChildren();
+  std::set_intersection(
+      AllChildrenInSubtree.begin(), AllChildrenInSubtree.end(),
+      OtherFeatures.begin(), OtherFeatures.end(),
+      std::inserter(Intersection, std::next(Intersection.begin())));
+  // if the children are not matching the other features, the intersection is
+  // empty and we are good to go for deletion
+  return std::optional<bool>{Intersection.begin() == Intersection.end()};
+}
+
+std::vector<detail::FeatureVariantTy> removeFeatures(
+    FeatureModel &FM, std::vector<detail::FeatureVariantTy>::iterator Begin,
+    std::vector<detail::FeatureVariantTy>::iterator End, bool Recursive) {
+  // if everything was deleted
+  std::vector<detail::FeatureVariantTy> NotDeletedFeatures;
+  if (Begin == End) {
+    return NotDeletedFeatures;
+  }
+
+  // 3 use-cases:
+  // If Feature is Leave --> can be deleted
+  // If Feature is not a Leave and !Recursive --> cannot be deleted
+  // If Recursive --> partition by "youngest" feature in sub-tree --> delete
+  // those in recursive mode
+  auto DeleteIt = std::partition(
+      Begin, End, [&FM, Recursive, &Begin, &End](detail::FeatureVariantTy &FV) {
+        if (fvIsLeave(FM, FV).value_or(false)) {
+          return true;
+        }
+        if (!Recursive) {
+          return false;
+        }
+        std::set<FeatureTreeNode *> OtherFeatures;
+        std::transform(
+            Begin, End,
+            std::inserter(OtherFeatures, std::next(OtherFeatures.begin())),
+            [&FM, &OtherFeatures](detail::FeatureVariantTy &FV) {
+              return getActualFeature(FM, FV);
+            });
+        OtherFeatures.erase(nullptr);
```

Comment:
```suggestion
        // Remove nullptr, in case, on of the specified Features could 
        // not be found in the FeatureModel 
        OtherFeatures.erase(nullptr);
```